swift-sdk/Internal/IterableAPIInternal.swift

```diff
@@ -0,0 +1,819 @@
+                                                  networkSession: URLSession(configuration: URLSessionConfiguration.default))
+        }
+        return _sharedInstance!
+    }
+
+    private func handle(launchOptions: [UIApplicationLaunchOptionsKey: Any]?) {
+        guard let launchOptions = launchOptions else {
+            return
+        }
+        if let remoteNotificationPayload = launchOptions[UIApplicationLaunchOptionsKey.remoteNotification] as? [AnyHashable : Any] {
+            if let _ = IterableUtil.rootViewController {
+                // we are ready
+                IterableAppIntegration.implementation?.performDefaultNotificationAction(remoteNotificationPayload)
+            } else {
+                // keywindow not set yet
+                DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
+                    IterableAppIntegration.implementation?.performDefaultNotificationAction(remoteNotificationPayload)
+                }
+            }
+        }
+    }
+
+    private func checkForDeferredDeeplink() {
+        guard config.checkForDeferredDeeplink else {
+            return
+        }
+        guard localStorage.ddlChecked == false else {
+            return
+        }
+        
+        guard let request = IterableRequestUtil.createPostRequest(forApiEndPoint: ITBConsts.linksEndpoint, path: ENDPOINT_DDL_MATCH, args: [ITBL_KEY_API_KEY : apiKey], body: DeviceInfo.createDeviceInfo()) else {
+            ITBError("Could not create request")
+            return
+        }
+        
+        NetworkHelper.sendRequest(request, usingSession: networkSession).observe {(result) in
+            switch result {
+            case .value(let json):
+                self.handleDDL(json: json)
+            case .error(let failureInfo):
+                if let errorMessage = failureInfo.errorMessage {
+                    ITBError(errorMessage)
+                }
+            }
+        }
+    }
+    
+    private func handleDDL(json: [AnyHashable : Any]) {
+        if let serverResponse = try? JSONDecoder().decode(ServerResponse.self, from: JSONSerialization.data(withJSONObject: json, options: [])),
+            serverResponse.isMatch,
+            let destinationUrlString = serverResponse.destinationUrl,
+            let action = IterableAction.actionOpenUrl(fromUrlString: destinationUrlString) {
+            let context = IterableActionContext(action: action, source: .universalLink)
+            
+            DispatchQueue.main.async {
+                IterableActionRunner.execute(action: action,
+                                             context: context,
+                                             urlHandler: IterableUtil.urlHandler(fromUrlDelegate: self.urlDelegate, inContext: context),
+                                             urlOpener: AppUrlOpener())
+            }
+        }
+        
+        localStorage.ddlChecked = true
+    }
+    
+    private func updateDeviceId() {
```

Comment:
What do you think about initializing deviceId lazily by moving this into the getter method?

Reply:
Not sure. Do we gain much be creating a separate property on IterableSDK. I don't think it matters much either way.

Reply:
On the other hand it is probably a little cleaner.